text/002x-succinct-atomic-swaps.md

```diff
@@ -0,0 +1,139 @@
+using Hash-based Time Locked Contracts in the [grinswap](https://github.com/GrinSwap/proof-of-concept) wallet project.
+
+# Motivation
+[motivation]: #motivation
+
+This RFC aims to show a succinct atomic swap protocol using adaptor signatures, accompanied by an implementation in the reference Mimblewimble wallet [grin-wallet](https://github.com/geneferneau/grin-wallet/tree/atomic). All Grin wallet implementations should be able to use this RFC to interoperate, and perform atomic swaps with each other.
+
+# Community-level explanation
+[community-level-explanation]: #community-level-explanation
+
+In an atomic swap, two parties agree to lock funds on one chain (e.g. Bitcoin), and interactively unlock funds on another chain (e.g. Grin). By unlocking funds on the Grin chain, the secret to unlock the funds on the Bitcoin chain is revealed. Revealing the secret this way allows the exchange to happen "atomically": when one lock is opened, so is the other lock. This atomic structure keeps one party from being able to steal funds from the other party: either both locks are opened, or neither are opened.
+
+The reference implementation uses Bitcoin as the other chain in the swap, but potentially any other chain using Secp256k1 keys and multisignatures can work with the protocol.
+
+# Reference-level explanation
+[reference-level-explanation]: #reference-level-explanation
+
+Funds on the non-Grin chain need to be locked in a multisignature transaction that requires knowledge of both "atomic nonces" to be spent. A refund transaction is also needed to recover funds in case one of the parties becomes non-responsive, or the protocol needs to be aborted for another reason.
+
+## Atomic nonces
+[atomic-nonces]: #atomic-nonces
+
+In the case of Grin, atomic nonces are just Secp256k1 private keys. Let's say Alice knows atomic nonce `aN`, and Bob knows atomic nonce `bN`.
+
+In the refund transaction, Alice (with locked Grin funds) will reveal her nonce `aN` to allow Bob to recover his locked funds on the other chain.
+
+In the main transaction, Bob and Alice create a multisignature Grin transaction which reveals Bob's nonce `bN` to Alice. Alice can then recover the funds locked on the other chain.
+
+Atomic IDs are identifiers used to derive atomic nonces using the keychain. Atomic IDs are formatted as follows:
+
+```
+0x4 | 'mwatomic' | 0x0 0x0 0x0 0x0 | 32-bit ID
+```
+
+The 32-bit ID can be expanded to 64-bits if necessary, and doesn't need to be random (it can be a simple counter). However, an ID can only be used once.
+
+Atomic nonces must only be used **once**. Reuse potentially results in loss of funds, if the other party has the previously used nonce.
+
+Without automated defenses, a malicious party could trick a user into deriving the same nonce (by supplying a duplicate atomic ID). The malicious user could then recover funds on the other chain without completing the atomic swap protocol.
```

Comment:
What existing defenses, if any, does the wallet have to avoid generating the same blinding factor for two outputs?

Reply:
Added some lines about how `key_id`s are derived for outputs, and subsequently used for blinding factors and signature nonces. Also explain why the strictly increasing counter **alone** won't work for atomic ID's.